src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -287,6 +316,84 @@ public TestResult findCorrespondingResult(String id) {
         return Collections.emptyList();
     }
 
+    /**
+     * Start for Test Result Trends.
+     * From here onwards start the code area responsible for generating various test result trends.
+     */
+
+    public hudson.tasks.junit.TestResult loadXml(){return null;}
+
+    /**
+     * A method for getting the list of packages for all levels of hierarchy.
+     * @return Array of packages of all hierarchies.
+     */
+    public String[] getProjectList(){
+
+        /*
+         * If project list is already ready no need to construct again.
+         */
+        if(projectList!=null) return projectList;
+        hudson.tasks.junit.TestResult r = loadXml();
+        Collection<SuiteResult> suiteList = r.getSuites();
+
+        /*
+         * A set for the package names.
+         */
+        Map<Integer,HashSet<String>> projectSet = new HashMap<Integer,HashSet<String>>();
+        int projectCount = 0;
+        int projectCountLimit = 50;
+        int levelCap = Integer.MAX_VALUE;
+        for(SuiteResult sr: suiteList){
+            for(CaseResult cr: sr.getCases()){
+                String caseName = cr.getFullName();
+                String projectName = "";
+
+                /*
+                 * Splitting package by treating "." as separator.
+                 */
+                String[] packageTree = caseName.split("[.]");
```

Comment:
The split should be at suite level, right ? Not at case level.

Reply:
By traversing over each case I am just extracting the package name. I don't think it will make a difference whether we iterate of testcases or testsuites for that purpose.